cmd/gardener-scheduler/app/gardener_scheduler.go

```diff
@@ -0,0 +1,373 @@
+func NewGardenerScheduler(cfg *config.SchedulerConfiguration) (*GardenerScheduler, error) {
+	if cfg == nil {
+		return nil, errors.New("config is required")
+	}
+
+	// validate the configuration
+	if err := validation.ValidateConfiguration(cfg); err != nil {
+		return nil, err
+	}
+
+	// Initialize logger
+	logger := logger.NewLogger(cfg.LogLevel)
+	logger.Info("Starting Gardener scheduler ...")
+
+	if flag := flag.Lookup("v"); flag != nil {
+		if err := flag.Value.Set(fmt.Sprintf("%d", cfg.KubernetesLogLevel)); err != nil {
+			return nil, err
+		}
+	}
+
+	// Prepare a Kubernetes client object for the Garden cluster which contains all the Clientsets
+	// that can be used to access the Kubernetes API.
+	if kubeconfig := os.Getenv("KUBECONFIG"); kubeconfig != "" {
+		cfg.ClientConnection.Kubeconfig = kubeconfig
+	}
+
+	restCfg, err := server.RestConfigFromClientConnectionConfiguration(cfg.ClientConnection)
+	if err != nil {
+		return nil, err
+	}
+
+	disc, err := discoveryFromSchedulerConfiguration(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	k8sGardenClient, err := kubernetes.NewForConfig(restCfg, client.Options{
+		Mapper: restmapper.NewDeferredDiscoveryRESTMapper(disc),
+		Scheme: kubernetes.GardenScheme,
+	})
+	if err != nil {
+		return nil, err
+	}
+	k8sGardenClientLeaderElection, err := k8s.NewForConfig(restCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up leader election if enabled and prepare event recorder.
+	var (
+		leaderElectionConfig *leaderelection.LeaderElectionConfig
+		recorder             = server.CreateRecorder(k8sGardenClient.Kubernetes(), "gardener-scheduler")
+	)
+	if cfg.LeaderElection.LeaderElect {
+		leaderElectionConfig, err = server.MakeLeaderElectionConfig(cfg.LeaderElection.LeaderElectionConfiguration, cfg.LeaderElection.LockObjectNamespace, cfg.LeaderElection.LockObjectName, k8sGardenClientLeaderElection, recorder)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &GardenerScheduler{
+		Config:                 cfg,
+		Logger:                 logger,
+		Recorder:               recorder,
+		K8sGardenClient:        k8sGardenClient,
+		K8sGardenInformers:     gardeninformers.NewSharedInformerFactory(k8sGardenClient.Garden(), 0),
+		K8sGardenCoreInformers: gardencoreinformers.NewSharedInformerFactory(k8sGardenClient.GardenCore(), 0),
+		KubeInformerFactory:    kubeinformers.NewSharedInformerFactory(k8sGardenClient.Kubernetes(), 0),
+		LeaderElection:         leaderElectionConfig,
+	}, nil
+}
+
+func (g *GardenerScheduler) cleanup() {
+	if err := os.RemoveAll(schedulerconfigv1alpha1.DefaultDiscoveryDir); err != nil {
+		g.Logger.Errorf("Could not cleanup base discovery cache directory: %v", err)
+	}
+}
+
+// Run runs the Gardener Scheduler. This should never exit.
+func (g *GardenerScheduler) Run(ctx context.Context, cancel context.CancelFunc) error {
+	defer g.cleanup()
+	leaderElectionCtx, leaderElectionCancel := context.WithCancel(context.Background())
```

Comment:
Please check whether you can `defer` some of the cancel functionality.

Reply:
add defer leaderElectionCancel() in the next Line to make sure the LeaderElection context is being terminated